text/0000-guaranteed-slice-repr.md

```diff
@@ -0,0 +1,182 @@
+      reason for this is that enum layout optimizations may rely on references
+      (including slices of any length) being aligned and non-null to distinguish
+      them from other data. You can obtain a pointer that is usable as `data`
+      for zero-length slices using [`NonNull::dangling()`].
+
+* `data` must point to `len` consecutive properly initialized values of type `T`.
+
+* The total size `len * mem::size_of::<T>()` of the slice must be no larger than `isize::MAX`,
+  and adding that size to `data` must not "wrap around" the address space.
+  See the safety documentation of [`pointer::offset`].
+
+## `str`
+
+The layout of `&str` is the same as that of `&[u8]`, and the layout of
+&mut str` is the same as that of `&mut [u8]`. More generally, `str` behaves like
+`#[repr(transparent)] struct str([u8]);`. Safe Rust functions may assume that
+`str` holds valid UTF8, but [it is not immediate undefined-behavior to store
+non-UTF8 data in `str`](https://doc.rust-lang.org/std/primitive.str.html#invariant).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Zero-sized types
+
+One could imagine representing `&[T]` as only `len` for zero-sized `T`.
+This proposal would preclude that choice in favor of a standard representation
+for slices regardless of the underlying type.
+
+Alternatively, we could choose to guarantee that the data pointer is present if
+and only if `size_of::<T> != 0`. This has the possibility of breaking exising
+code which smuggles pointers through the `data` value in `from_raw_parts` /
+`into_raw_parts`.
+
+## Uninhabited types
+
+Similarly, we could be *extra* tricky and make `&[!]` or other `&[Uninhabited]`
+types into a ZST since the slice can only ever be length zero. This may offer
+modest performance benefits for highly generic code which happens to create
+empty slices of uninhabited types, but this is unlikely to be worth the
+cost of maintaining a special case.
+
+## Compatibility with C++ `std::span`
+
+The largest drawback of this layout and set of validity requirements is that it
+may preclude `&[T]` from being representationally equivalent to C++'s
+`std::span<T, std::dynamic_extent>`.
+
+* `std::span` does not currently guarantee its layout. In practice, pointer + length
+  is the common representation. This is even observable using `is_layout_compatible`
+  [on MSVC](https://godbolt.org/z/Y8ardrshY), though not
+  [on GCC](https://godbolt.org/z/s4v4xehnG) nor
+  [on Clang](https://godbolt.org/z/qsd1K5oGq). Future changes to guarantee a
+  different layout in the C++ standard (unlikely due to MSVC ABI stabilitiy
+  requirements) could preclude matching the layout with `&[T]`.
+
+* Unlike Rust, `std::span` allows the `data` pointer to be `nullptr`. One
+  possibile workaround for this would be to guarantee that `Option<&[T]>` uses
+  `data: std::ptr::null()` to represent the `None` case, making `std::span<T>`
+  equivalent to `Option<&[T]>` for non-zero-sized types.
```

Comment:
We would probably also need to guarantee that the length part of the pointer is 0 (or at least not uninitialized) when `Option<&[T]>` is `None`. Note that this is not currently the case, `None::<&[u8]>` is currently represented as `[nullptr, uninit]` (though this is not guaranteed).